File: samples/Showcase.cs
// Godot Classic color showcase (C#) — illustrative, not functional code.
// Shows every color the C# grammar can emit.
using System.Collections.Generic;
using Godot;

namespace Game.Entities;

#region Types
public enum State { Idle, Running, Dead }
#endregion

/// <summary>
/// A player with health and movement.
/// </summary>
[GlobalClass]
public partial class Player : CharacterBody2D
{
    private const int MaxHealth = 100;

    [Signal] public delegate void ExampleSignalEventHandler(int oldValue, int newValue);
    [Export] public float Speed { get; set; } = 220.5f;

    private State _state = State.Idle;
    private int _health = MaxHealth;
    private bool _alive = true;
    private Node2D _target = null;
    private readonly List<string> _inventory = new() { "sword", "shield" };

    public void ExampleFunction(int amount)
    {
        // Clamp health, then notify listeners
        int previous = _health;
        _health = Mathf.Max(0, _health - amount);
        EmitSignal(SignalName.ExampleSignal, previous, _health);

        if (_health <= 0)
        {
            _alive = false;
            _state = State.Dead;
            return;
        }
        else if (_health < 25)
            GD.Print($"Low health: {_health}\n");

#if DEBUG
        for (int i = 0; i < _inventory.Count; i++)
            GD.Print($"slot {i}: {_inventory[i]}\t");
#endif
    }
}

File: samples/showcase.rs
//! Godot Classic color showcase (godot-rust) — illustrative, not functional code.
use godot::classes::CharacterBody2D;
use godot::prelude::*;

const MAX_HEALTH: i32 = 100;

/// A player with health and movement.
#[derive(GodotClass)]
#[class(base = CharacterBody2D)]
pub struct Player {
    #[export]
    speed: f32,
    #[var]
    health: i32,
    alive: bool,
    inventory: Vec<GString>,
    base: Base<CharacterBody2D>,
}

#[godot_api]
impl Player {
    #[signal]
    fn example_signal(old_value: i32, new_value: i32);

    #[func]
    fn example_function(&mut self, amount: i32) {
        // Clamp health, then notify listeners
        let previous = self.health;
        self.health = (self.health - amount).max(0);
        self.signals().example_signal().emit(previous, self.health);

        if self.health <= 0 {
            self.alive = false;
            godot_warn!("Player died\n");
            return;
        }

        for (index, item) in self.inventory.iter().enumerate() {
            godot_print!("slot {index}: {item}\t");
        }
    }
}
